mox/impl/src/lib.rs

```diff
@@ -1,245 +1,239 @@
-    fn to_tokens(&self, tokens: &mut TokenStream) {
-        tag_to_tokens(&self.name, &self.attributes, Some(&self.children), tokens);
+fn mangle_expr_path(name: &mut syn::ExprPath) {
+    for segment in name.path.segments.iter_mut() {
+        mangle_ident(&mut segment.ident);
     }
 }
 
-fn tag_to_tokens(
-    name: &Ident,
-    attributes: &[MoxAttr],
-    children: Option<&[MoxItem]>,
-    stream: &mut TokenStream,
-) {
-    // this needs to be nested within other token groups, must be accumulated
-    // separately from stream
-    let mut contents = quote!();
+impl TryFrom<syn_rsx::Node> for MoxAttr {
+    type Error = syn::parse::Error;
 
-    for attr in attributes {
-        attr.to_tokens(&mut contents);
-    }
-
-    if let Some(items) = children {
-        for child in items {
-            MoxAttr::child(child).to_tokens(&mut contents);
+    fn try_from(node: syn_rsx::Node) -> syn::Result<Self> {
+        use syn_rsx::NodeType::*;
+        match node.node_type {
+            Element | Text | Block => Err(node_parse_error(&node, "MoxAttr")),
+            Attribute => {
+                Ok(MoxAttr { name: MoxAttr::validate_name(node.name.unwrap())?, value: node.value })
+            }
         }
     }
-
-    quote!(mox::topo::call(|| { #name() #contents .build() })).to_tokens(stream);
 }
 
-impl From<SnaxTag> for MoxTag {
-    fn from(SnaxTag { name, attributes, children }: SnaxTag) -> Self {
-        let attributes = attributes.into_iter().map(MoxAttr::from).collect();
-        Self {
-            // TODO get the span for the whole tag
-            span: name.span(),
-            name,
-            attributes,
-            children: children.into_iter().map(MoxItem::from).collect(),
+impl MoxAttr {
+    fn validate_name(name: syn_rsx::NodeName) -> syn::Result<syn::Ident> {
+        use syn::{parse::Error, punctuated::Pair, PathSegment};
+        use syn_rsx::NodeName;
+
+        let invalid_error = |span| Error::new(span, "Invalid name for an attribute");
+
+        match name {
+            NodeName::Path(syn::ExprPath {
+                attrs,
+                qself: None,
+                path: syn::Path { leading_colon: None, mut segments },
+            }) if attrs.is_empty() && segments.len() == 1 => {
+                let pair = segments.pop();
+                match pair {
+                    Some(Pair::End(PathSegment { mut ident, arguments }))
+                        if arguments.is_empty() =>
+                    {
+                        mangle_ident(&mut ident);
+                        Ok(ident)
+                    }
+                    _ => Err(invalid_error(segments.span())),
+                }
+            }
+            NodeName::Dash(punctuated) => {
+                Err(Error::new(punctuated.span(), "Dash attribute name syntax isn't supported"))
+            }
+            NodeName::Colon(punctuated) => {
+                Err(Error::new(punctuated.span(), "Colon attribute name syntax isn't supported"))
```

Comment:
Is this possible to support? Our tag names desugar to Rust function calls which could be at a nested module path. It could make the interaction between mox tags and "normal" rust code much clearer.

Reply:
It is an attribute name, which becomes name for a method call, which i don't see anything but an `Ident`. Maybe we could do something like `namespace::Into<T>:into = [value]` -> `<Thing as Into<T>>::into(thing, [value])` or whatever but that's another bag of worms.

Reply:
This is actually a thing now for the tag names tho.

Reply:
This all makes sense. Also I'll note for myself that after reading more about how syn_rsx works, I'm pretty sure this colon variant is for a single colon separating them, as opposed to the path.